File: SOC/src/cc/game/soc/core/Tile.java
package cc.game.soc.core;

import java.util.List;

import cc.lib.game.IVector2D;
import cc.lib.game.Utils;
import cc.lib.utils.Reflector;


/**
 * 
 * @author Chris Caron
 * 
 */
public final class Tile extends Reflector<Tile> implements IVector2D {
	
	static {
		addAllFields(Tile.class);
	}
	
	//	 position of cell
	private float	x, y;

	//	 indicies to vertices
	// that are adjacent to
	// this cell
	private int[] adjVerts	= new int[6];
	private int	numAdj;
	private int	dieNum;
	private int islandNum;
	private TileType type = TileType.NONE;
	private int pirateRouteNext = -1; // When 
	
	public Tile() {}
	
	/**
	 * 
	 * @param x
	 * @param y
	 */
	Tile(float x, float y, TileType type) {
		this.x = x;
		this.y = y;
		this.type = type;
	}

	void reset() {
	}

	/**
	 * @return Returns the num.
	 */
	public int getDieNum() {
		return dieNum;
	}
	
	/**
	 * 
	 * @param dieNum
	 */
	public void setDieNum(int dieNum) {
		this.dieNum = dieNum;
	}

	/**
	 * @return Returns the type.
	 */
	public TileType getType() {
		return type;
	}

	/**
	 * 
	 * @param type
	 */
	public void setType(TileType type) {
		this.type = type;
	}
	
	/**
	 * 
	 * @param type
	 * @param dieNum
	 */
	public void setType(TileType type, int dieNum) {
		this.type = type;
		this.dieNum = dieNum;
	}
	
	/**
	 * @return Returns the x.
	 */
	public float getX() {
		return x;
	}

	/**
	 * @return Returns the y.
	 */
	public float getY() {
		return y;
	}

	/**
	 * @return Returns the adjVerts.
	 */
	public List<Integer> getAdjVerts() {
		return Utils.asList(adjVerts, 0, numAdj);
	}

	/**
	 * @return Returns the numAdj.
	 */
	public int getNumAdj() {
		return numAdj;
	}
	
	public int getAdjVert(int index) {
		return adjVerts[index];
	}

    void setAdjVerts(int[] adjVerts) {
        assert(adjVerts.length == 6);
        this.adjVerts = adjVerts;
        this.numAdj = adjVerts.length;
    }
    
    void setX(float x) {
    	this.x = x;
    }
	
    void setY(float y) {
    	this.y = y;
    }

	public boolean isWater() {
		return type.isWater;
	}
	
	public boolean isLand() {
		return type.isLand;
	}
	
	public ResourceType getResource() {
		return getType().resource;
	}
	
	public CommodityType getCommodity() {
		return getType().commodity;
	}

	public final int getIslandNum() {
		return islandNum;
	}

	public final void setIslandNum(int islandNum) {
		this.islandNum = islandNum;
	}

	public final boolean isDistributionTile() {
		return getType().isDistribution;
	}
	
	public boolean isPort() {
		return getType().isPort;
	}
	
	@Override
	public String toString() {
		String r = type.name();
		if (isPort())
			r += " Port";
		if (islandNum > 0)
			r +=  "isle[" + islandNum + "]";
		if (dieNum > 0)
			r += " Die[" + dieNum + "]";
		return r;
	}
	
	@Override
	public boolean equals(Object o) {
		Tile t = (Tile)o;
		return t.getX() == getX() && t.getY() == getY();
	}

	public final int getPirateRouteNext() {
		return pirateRouteNext;
	}

	public final void setPirateRouteNext(int pirateRouteNext) {
		this.pirateRouteNext = pirateRouteNext;
	}

}

File: SOC/src/cc/game/soc/core/Tile.kt
package cc.game.soc.core

import cc.lib.game.IVector2D
import cc.lib.game.Utils
import cc.lib.reflector.Reflector

/**
 *
 * @author Chris Caron
 */
class Tile : Reflector<Tile>, IVector2D {
	companion object {
		init {
			addAllFields(Tile::class.java)
		}
	}

	//	 position of cell
	private var x = 0f
	private var y = 0f

	//	 indicies to vertices
	// that are adjacent to
	// this cell
    @JvmField
    var adjVerts = IntArray(6)

	/**
	 * @return Returns the numAdj.
	 */
	var numAdj = 0
		private set
	/**
	 * @return Returns the num.
	 */
	/**
	 *
	 * @param dieNum
	 */
	var dieNum = 0
	var islandNum = 0
	/**
	 * @return Returns the type.
	 */
	/**
	 *
	 * @param type
	 */
	var type = TileType.NONE
	var pirateRouteNext = -1 // When

	constructor() {}

	/**
	 *
	 * @param x
	 * @param y
	 */
	internal constructor(x: Float, y: Float, type: TileType) {
		this.x = x
		this.y = y
		this.type = type
	}

	fun reset() {}

	/**
	 *
	 * @param type
	 * @param dieNum
	 */
	fun setType(type: TileType, dieNum: Int) {
		this.type = type
		this.dieNum = dieNum
	}

	/**
	 * @return Returns the x.
	 */
	override fun getX(): Float {
		return x
	}

	/**
	 * @return Returns the y.
	 */
	override fun getY(): Float {
		return y
	}

	/**
	 * @return Returns the adjVerts.
	 */
	fun getAdjVerts(): List<Int> {
		return Utils.asList(adjVerts, 0, numAdj)
	}

	fun getAdjVert(index: Int): Int {
		return adjVerts[index]
	}

	fun setAdjVerts(adjVerts: IntArray) {
		assert(adjVerts.size == 6)
		this.adjVerts = adjVerts
		numAdj = adjVerts.size
	}

	fun setX(x: Float) {
		this.x = x
	}

	fun setY(y: Float) {
		this.y = y
	}

	val isWater: Boolean
		get() = type.isWater
	val isLand: Boolean
		get() = type.isLand
	val resource: ResourceType?
		get() = type.resource
	val commodity: CommodityType?
		get() = type.commodity
	val isDistributionTile: Boolean
		get() = type.isDistribution
	val isPort: Boolean
		get() = type.isPort

	override fun toString(): String {
		var r = type.name
		if (isPort) r += " Port"
		if (islandNum > 0) r += "Isle[$islandNum]"
		if (dieNum > 0) r += " Die[$dieNum]"
		return r
	}

	override fun equals(o: Any?): Boolean {
		val t = o as Tile?
		return t!!.getX() == getX() && t.getY() == getY()
	}
}